# phyloscanner
Analysing phylogenetic diversity and relationships between and within hosts at once, in windows along the genome.  
Dependencies: [samtools](http://www.htslib.org/), [pysam](https://github.com/pysam-developers/pysam), [biopython](http://biopython.org/wiki/Download), [mafft](http://mafft.cbrc.jp/alignment/software/) and [RAxML](http://sco.h-its.org/exelixis/web/software/raxml/index.html).  

<p align="center"><img src="InfoAndInputs/PhyloscannerDiagram_big2_ShiverNotNamed.jpg" alt="Phyloscanner" width="820", height="315"/></p>

The input to phyloscanner is mapped reads - bam files.
We wrote phyloscanner to analyse bam files that each represent a pathogen population (HIV, specifically) in one host, exhibiting within-host and between-host diversity; in general use each bam file should be a sample representing some subpopulation, and we analyse within- and between-sample diversity.

### Basic usage:
With the initial `$` traditionally indicating that we're on the command line, the basic command looks like
```bash
$ phyloscanner.py bams.txt refs.txt --windows 1,300,301,600,...
```
where  
1. `bams.txt` is a plain text file listing the desired bam files, one per line;  
2. `refs.txt` is a plain text file listing the files containing the sequences to which the reads were mapped in order to create the bam files (the *references*), one per bam file, in the same order as in `bams.txt`;  
3. the `--windows` option is used to specify an even number of comma-separated positive integers: these are the coordinates of the windows to analyse, interpreted pairwise, i.e. the first two are the left and right edges of the first window, the third and fourth are the left and right edges of the second window, ... i.e. in the above example we have windows 1-300, 301-600, ...  

Details of optional arguments of phyloscanner can be seen by using the `--help` option, or seeing the section *Phyloscanner options* below.

### An Example

The simulated bam files in [ExampleInputData](ExampleInputData) illustrate interesting within- and between-host diversity.
Try running this command from the directory where your local copy of this code repostory lives, or adjusting paths appropriately if run from another directory:
```bash
# First make a list of the desired bams and the desired refs, in matching order:
$ ls ExampleInputData/*.bam > bams.txt
$ ls ExampleInputData/*_ref.fasta > refs.txt
# Then
$ ./phyloscanner.py bams.txt refs.txt --windows 1000,1300,2000,2300,3000,3300,4000,4300,5000,5300,6000,6300,7000,7300,8000,8300 -A InfoAndInputs/2refs_HXB2_C.BW.fasta --pairwise-align-to B.FR.83.HXB2_LAI_IIIB_BRU.K03455 
```
(Those windows make best use of this simulated data, the `-A` option includes an alignment of extra reference sequences along with the reads, and the `--pairwise-align-to` option is explained later on.)
In the output you'll see trees and summary information indicating that these samples constitute:
* a singly infected individual, 
* a dually infected individual i.e. infected by two distinct strains of virus,
* a contamination pair (one contaminating the other), and
* a pair exhibiting *ancestry*, i.e. one having evolved from the other. For populations of pathogens, this implies transmission, either indirectly (via unsampled intermediate patients) or directly.

Here is one of the trees from the output.
Each patient has many sequences (reads), with one colour per patient.
Extra reference sequences and any reads phyloscanner decides it should ignore are coloured black.
Can you see why each of the patients merits their label?

<p align="center"><img src="InfoAndInputs/Tree_SimulatedData_InWindow_4000_to_4300.jpg" alt="ExampleTree" width="650", height="741"/></p>

(These bam files were generated by taking HIV sequences from the [LANL HIV database](https://www.hiv.lanl.gov/content/sequence/NEWALIGN/align.html), using them as starting points for simulated evolution (performed using [SeqGen](https://github.com/rambaut/Seq-Gen)) calibrated to our own real data on within-host diversity.
Reads were then generated in eight windows of the genome - those specified with the `--windows` option in the command above - instead of all along the genome, purely to keep file sizes nice and small while remaining an interesting example.)


### What windows should I choose for my own data?
#### TODO: revisit this section in light of overlaps etc.
I'm glad you asked. It's important. You might as well fully cover the genomic region you're interested in. That requires choosing where to start and where to end. If you're interested in the whole genome, the start is 1 and the end is the genome length, or more precisely the length of an alignment of the references in the bam files (this alignment is generated by phyloscanner, and it is with respect to this alignment that coordinates are interpreted by default; more on this later).  
If your reads were generated by amplifying the sample using primers, the primers should have been trimmed from the reads as part of whatever bioinformatic pipeline produced your input bam files. (This can be done for example using [Fastaq](https://github.com/sanger-pathogens/Fastaq), which is called as part of the [shiver](https://github.com/ChrisHIV/shiver) pipeline.)
Then a sensible choice for the start of the first window would be the first position *after* the first primer, and a sensible choice for the end of the last window would be the last position *before* the last primer.

In addition, you need to choose how wide each window is, and how much neighbouring windows overlap (with negative overlap understood to mean that there is space in between neighbouring windows). These are a bit more complicated.

Window width first. If a window is very small, so little diversity is contained inside it (within or between samples) that the number of *unique* reads overlapping the window is small, hindering meaningful phylogenetics. If window width exceeds read length, then you will have no reads in the window, since we keep only reads fully overlapping the window. Somewhere between these two extremes therefore maximises the number of unique reads; to help you figure out what that is for your data, you can run phyloscanner with the `--explore-window-widths` option. This reports, for each in a list of window widths to try, how many unique reads are found for each bam and at each position along the genome. To summarise these read counts into a single value that varies with window width, you could use the mean, or the median; or you might be interested in a percentile lower than the 50th, if your concern is ensuring some minimal amount of diversity across all bams and all genomic positions. Up to you.  
(Note that some of the other options can affect how many reads you get in a window and so can affect what `--explore-window-widths` will tell you, namely the `--excision-coords`, `--merging-threshold`, `--min-read-count`, `--quality-trim-ends`, `--min-internal-quality`, and `--discard-improper-pairs` options. The first two can result in two or more unique reads being merged into one; the rest can simply discard some reads. You could choose values for the associated parameters immediately and then use `--explore-window-widths`, or else come back to `--explore-window-widths` later on once you've got the hang of phyloscanner and investigated the effect of those other options in your data.)  
NB power users might want to optimise their own measure of phylogenetic information as a function of window width; one of the first metrics to pop into your head might be the mean bootstrap of all nodes in the tree. That's not advised because within a sample there may be many very similar sequences, and the set of nodes connecting these may have poor boostrap support, but this is not something that ought to be penalised. Also in theory you might be able to increase the window width until only a single read is found spanning the window in each patient; your bootstraps might then be great - between-host diversity is greater than that within-host - but you've thrown out all the within-host information.

Now overlap. It's probably a good idea to choose the overlap such that each window is independent of its neighbours, because if you have sliding windows that only shift by one base each time, you're redoing calculations and analysis for virtually the same data again and again. Independent windows can be achieved trivially by having zero overlap or negative overlap (i.e. space) between neighbouring windows. But you might want windows to overlap so that you can fit more windows into the same space. In that case, independency of windows can be achieved non-trivially by choosing the overlap such that the distance from the start of one window to the end of the next exceeds the read length: that way no read can span two windows and get counted twice. This would be unambiguous if you only had one bam file, or if every bam file used exactly the same reference for mapping. With multiple references however, window coordinates need to be interpreted with respect to different references. If one reference has a deletion inside a particular window, then the window width is smaller for that reference for that window. So, if your window width & overlap is such that reads are *only just unable* to fully span two neighbouring windows, a small deletion could make some reads *only just able* to fully span two neighbouring windows. (The fraction of such reads will be small if coverage is uniform around the two windows.) If you decrease the overlap so that the distance from the start of one window to the end of the next window is somewhat larger than read length, a correspondingly larger deletion would be required inside those two windows to allow any reads to fully span them. However decreasing the overlap also decreases the total number of windows you can fit in. To balance this you have to decide how much you care if, for one of many bam files and for one of many positions along the genome, one of many reads overlaps two neighbouring windows and so provides sequence data to both.

NB wherever *read* and *read length* appeared in the discussion above, they should be substituted for *insert* and *insert size* if you have paired-read data AND the reads in a pair sometimes overlap AND you run phyloscanner with `--merge-paired-reads` to merge overlapping paired reads into a single longer read (see the cartoon below). A complication with this is that whereas read length is typically fixed within a sample, insert size has a distribution of different values. A window which is wider than twice the read length can never get any reads, because the reads in a pair need to overlap in order to be merged. So you have two choices.  
1. Choose  
(read length) < (window width) < (twice the read length)  
Then you're restricted to the subset of read pairs that satisfy  
(window width) <= (insert size) < (twice the read length)  
because only such pairs can overlap and fully span the window. The fraction of such reads in a sample is the integral of the unit-normalised insert size distribution between the two limits in the inequality above.  
2. Choose  
(window width) <= (read length)  
Then you can have single reads contribute in addition to merged overlapping read pairs. But perhaps that window is too short; see the window-width discussion above.


### Interpreting window coordinates
By default the references used for mapping (to produce the bam files), together with an extra set of references if specified with `--alignment-of-other-refs`, are all aligned together and window coordinates are interpreted with respect to the alignment (i.e. position *n* refers to the *n*th column of that alignment, which could be a gap for some of the sequences). This alignment can be found in the file `RefsAln.fasta` after running phyloscanner, should you want to inspect it and possibly run again. You can manually specify window coordinates with respect to this alignment, using the `--windows` option, or have windows automatically chosen using `--auto-window-params`, which attempts to minimise the affect of insertions and deletions in the references on your window width and overlap preferences.  
Alternatively, if you are using the `--alignment-of-other-refs` option to include extra references, you can use `--pairwise-align-to` to name one of these references to be a kind of *reference reference*: instead of aligning of all the bam file references to each other, they will be sequentially and separately pairwise-aligned to your named reference, and window coordinates are interpreted with respect to that named reference.
Using the `--pairwise-align-to` option is expected to more stable than `--windows` or `--auto-window-params` if your bam file references are many and diverse, since pairwise alignment is easier than multiple sequence alignment.
It also has the advantage that when running phyloscanner more than once with different bam files, the coordinates mean the same thing each time.

### Some more pictures
Here are some cartoons of the `--merge-paired-reads`, `--merging-threshold` and `--excision-coords` options in action:

<p align="center"><img src="InfoAndInputs/OptionsDiagram.jpg" alt="Phyloscanner" width="750" height="377"/></p>

### Phyloscanner options
```
Window options - you must choose exactly one of -W, -AW or -E:
  -W WINDOWS, --windows WINDOWS
                        A comma-separated series of paired coordinates
                        defining the boundaries of the windows. e.g.
                        1,300,11,310,21,320 would define windows 1-300,
                        11-310, 21-320.
  -AW AUTO_WINDOW_PARAMS, --auto-window-params AUTO_WINDOW_PARAMS
                        Used to specify 2, 3 or 4 comma-separated integers
                        controlling the automatic creation of regular windows.
                        The first integer is the width you want windows to be,
                        weighting each column in the alignment of bam file
                        references (plus any extra references included) by its
                        non-gap fraction (so that windows become wider to
                        accommodate insertions). The second is the overlap
                        between the end of one window and the start of the
                        next (which can be negative, implying unused space in
                        between windows). The optional third integer is the
                        start position for the first window (by default, 1).
                        The optional fourth integer is the end position for
                        the last window (by default, windows will continue up
                        to the end of the alignment of references).
  -E EXPLORE_WINDOW_WIDTHS, --explore-window-widths EXPLORE_WINDOW_WIDTHS
                        Use this option to explore how the number of unique
                        reads found in each bam file in each window, all along
                        the genome, depends on the window width. After this
                        option specify a comma-separated list of integers. The
                        first integer is the starting point for stepping along
                        the genome, in case you're not interested in the very
                        beginning. The second integer is the width spanned by
                        two neighbouring windows; this sets the overlap for
                        each window width though the equation (width of two
                        windows) = 2x(window width) - (overlap between
                        windows), with negative overlap understood to mean
                        space between the end of one window and the start of
                        the next. Subsequent integers are window widths to
                        try. For example, if you specified
                        1000,301,101,151,201 we would count the number of
                        unique reads in windows 1000-1100, 1200-1300,
                        1400-1500, ... and in 1000-1150, 1150-1300, 1300-1450
                        ... and in 1000-1200, 1100-1300, 1200-1400, ... where
                        the dots denote continuation to the end of the genome.
                        (Note that both end coordinates of the window are
                        included, hence why 1000-1100 has width 101.) Output
                        is written to the file specified with the --explore-
                        window-width-file option.
  -EF EXPLORE_WINDOW_WIDTH_FILE, --explore-window-width-file EXPLORE_WINDOW_WIDTH_FILE
                        Used to specify an output file for window width data,
                        when the --explore-window-widths option is used.
                        Output is in in csv format.

Options we particularly recommend:
  -A ALIGNMENT_OF_OTHER_REFS, --alignment-of-other-refs ALIGNMENT_OF_OTHER_REFS
                        Used to specify an alignment of reference sequences
                        (which need not be those used to produce the bam
                        files) that will be cut into the same windows as the
                        bam files and included in the alignment of reads, for
                        comparison. This is required if phyloscanner is to
                        analyse the trees it produces.
  -RR REF_FOR_ROOTING, --ref-for-rooting REF_FOR_ROOTING
                        Used to name a reference sequence, which must be
                        present in the file you specify with -A, to be an
                        outgroup in each tree. This is required if
                        phyloscanner is to analyse the trees it produces.
  -2 PAIRWISE_ALIGN_TO, --pairwise-align-to PAIRWISE_ALIGN_TO
                        By default, phyloscanner figures out where
                        corresponding windows are in different bam files by
                        creating a multiple sequence alignment containing all
                        of the mapping references used to create the bam files
                        (plus any extra references included with -A), and
                        window coordinates are intepreted with respect to this
                        alignment. However using this option, the mapping
                        references used to create the bam files are each
                        separately pairwise aligned to one of the extra
                        references included with -A, and window coordinates
                        are interpreted with respect to this reference.

Options intended to minimise the impact of poor quality reads:
  -I, --discard-improper-pairs
                        For paired-read data, discard all reads that are
                        improperly paired: in the wrong orientation, or one
                        mate unmapped, or too far apart (as flagged at the
                        time of mapping).
  -Q1 QUALITY_TRIM_ENDS, --quality-trim-ends QUALITY_TRIM_ENDS
                        Each end of the read is trimmed inwards until a base
                        of this quality is met.
  -Q2 MIN_INTERNAL_QUALITY, --min-internal-quality MIN_INTERNAL_QUALITY
                        Discard reads containing more than one base of a
                        quality below this parameter. (If used in conjuction
                        with the --quality-trim-ends option, the trimming of
                        the ends is done first.)
  -MC MIN_READ_COUNT, --min-read-count MIN_READ_COUNT
                        Reads with a count less than this value (after
                        merging, if merging is being done) are discarded. The
                        default value of 1 means all reads are kept. You may
                        want to discard rare reads to protect against
                        sequencing error and/or low-level contamination.
                        Retaining fewer reads will also speed up all
                        subsequent processing and analysis of the reads.

Other assorted options:
  -AR, --allow-read-repeats
                        Some background first: by default, if a read with the
                        same name is found to span each of a series of
                        consecutive, overlapping windows, it is only used in
                        the first window. 'Consecutive' means next to each
                        other in the order you specified. For example, if you
                        specified windows 10-20, 15-25, 20-30 and 31-40, and
                        there was a read that spanned all four windows (i.e.
                        it started at or before position 10 and ended at or
                        after position 40), it would be used in window 10-20,
                        not used in 15-25 because it spanned the last window,
                        not used in 20-30 because it spanned the last window
                        (even though it was skipped there), and used in 31-40
                        because this window does not overlap with the last
                        one. This default behaviour allows windows to overlap
                        yet contain independent read data, which is good
                        because overlapping windows means more windows can be
                        fit into the same space. With paired read data, mates
                        in a pair have the same name; the default behaviour
                        without the --merge-paired-reads option is to use at
                        most one of the two mates (in consecutive overlapping
                        windows and in the same window), and the default
                        behaviour with the --merge-paired-reads option is to
                        merge mates into a single read, using this merged read
                        only the first time it is encountered in consecutive
                        overlapping windows. Using this option, --allow-read-
                        repeats, when reads are encountered again in
                        overlapping windows they will be used again, not
                        skipped. This means neighbouring overlapping windows
                        can no longer be considered independent.
  -C CONTAMINANT_COUNT_RATIO, --contaminant-count-ratio CONTAMINANT_COUNT_RATIO
                        Used to specify a numerical value which is interpreted
                        in the following 'way: if a sequence is found exactly
                        duplicated between any two bam files, 'and is more
                        common in one than the other by a factor at least
                        equal to this 'value, the rarer sequence is diagnosed
                        as contamination. It does not go into ' the tree, and
                        instead goes into a contaminant read fasta file.
  -CD CONTAMINANT_READ_DIR, --contaminant-read-dir CONTAMINANT_READ_DIR
                        A directory containing the contaminant read fasta
                        files produced by a previous run of phyloscanner.py
                        using the the -C flag: reads flagged as contaminants
                        there will be considered contaminants in this run too.
                        The windows must exactly match up between these two
                        runs, most easily achieved with the -2 option. The
                        point of this option is to first do a run with every
                        bam file you have in which there could conceivably be
                        cross-contamination, using the -C flag (and possibly
                        -CO to save time), and then in subsequent runs
                        focussing on subsets of bam files you will be able to
                        identify contamination from outside that subset.
  -F RENAMING_FILE, --renaming-file RENAMING_FILE
                        Specify a file with one line per bam file, showing how
                        reads from that bam file should be named in the output
                        files. (By default, each bam file's basename is used.)
  -MT MERGING_THRESHOLD, --merging-threshold MERGING_THRESHOLD
                        Reads that differ by a number of bases equal to or
                        less than this are merged, those with higher counts
                        effectively absorbing those with lower counts. The
                        default value of 0 means there is no merging.
  -N NUM_BOOTSTRAPS, --num-bootstraps NUM_BOOTSTRAPS
                        The number of bootstraps to be calculated for RAxML
                        trees (by default, none i.e. only the ML tree is
                        calculated).
  -Ns BOOTSTRAP_SEED, --bootstrap-seed BOOTSTRAP_SEED
                        The random-number seed for running RAxML with
                        bootstraps. The default is 1.
  -O, --keep-overhangs  Keep the part of the read that overhangs the edge of
                        the window. (By default this is trimmed, i.e. only the
                        part of the read inside the window is kept.) Keeping
                        overhangs means that, within each bam file, reads that
                        are identical inside the window but have different
                        overhangs will not be merged into a single sequence
                        (with a count greater than 1). Differences in
                        overhangs may be SNPs, or simply because the overhangs
                        start or end at different points; this option is
                        therefore a bit weird, because it's nice to merge all
                        reads that are identical inside the window of
                        interest.
  -P, --merge-paired-reads
                        For paired-read data for which the reads in a pair
                        (sometimes) overlap with each other: merge overlapping
                        paired reads into a single read. Allows wider windows
                        to be used.
  -RC REF_FOR_COORDS, --ref-for-coords REF_FOR_COORDS
                        (Deprecated; the --pairwise-align-to option is
                        expected to perform better.) If the --pairwise-align-
                        to option is not used, then a multiple sequence
                        alignment is created with all the mapping references
                        used to create the bam files (plus any extra
                        references included with -A). By default, window
                        coordinates are interpreted with respect to this
                        alignment, i.e. they are in the alignment coordinates.
                        With this option (--ref-for-coords), the multiple
                        sequence alignment is still created but window
                        coordinates are interpreted with respect to a named
                        reference, which must be one of those included with
                        -A.
  -XC EXCISION_COORDS, --excision-coords EXCISION_COORDS
                        Used to specify a comma-separated set of integer
                        coordinates that will be excised from the aligned
                        reads. Useful for sites of non-neutral evolution,
                        which distort phylogenies. Requires the -XR flag.
  -XR EXCISION_REF, --excision-ref EXCISION_REF
                        Used to specify the name of a reference (which must be
                        present in the file you specify with -A) with respect
                        to which the coordinates specified with -XC are
                        interpreted. If you are also using the --pairwise-
                        align-to option, you must use the same reference there
                        and here.
  --output-dir OUTPUT_DIR
                        Used to specify the name of a directory (which should
                        not exist already) in which all intermediate and
                        output files will be created.
  --time                Prints the times taken by different steps.
  --x-raxml X_RAXML     The command required to invoke RAxML. You may include
                        RAxML options in this command, which need to separated
                        by white space as usual and then the whole thing needs
                        to be surrounded with one pair of quotation marks (so
                        that the raxml command and its options are kept
                        together as one option for phyloscanner). If you
                        include a path to your raxml binary (necessary if it
                        is not in the $PATH variable of your terminal), it may
                        not include whitespace, since whitespace is
                        interpreted as separating raxml options. The default
                        is 'raxmlHPC-AVX -m GTRCAT -p 1', where -m specifies
                        an evolutionary model and -p specifies a random number
                        seed for the parsimony inferences. If changing from
                        the default, note that the -m and -p options are
                        compulsory. Do not include in this command options
                        relating to bootstraps: use phyloscanner's --num-
                        bootstraps and --bootstrap-seed options instead. Do
                        not include options relating to the naming of files.
  --x-mafft X_MAFFT     The command required to invoke mafft (by default:
                        mafft).
  --x-samtools X_SAMTOOLS
                        The command required to invoke samtools, if needed (by
                        default: samtools).

Options for detailed bioinformatic interrogation of the input bam files (not intended for normal usage):
  -IO, --inspect-disagreeing-overlaps
                        When read pairs are merged, those pairs that overlap
                        but disagree are discarded. With this option, these
                        discarded pairs are written to a bam file (one per
                        patient, with their reference file copied to the
                        working directory) for your inspection.
  -RN1, --read-names-1  Produce a file for each window and each bam, listing
                        the names (as they appear in the input bam file) of
                        the reads that phyloscanner used. If you like this you
                        may also like tools/ExtractNamedReadsFromBam.py, which
                        is run separately from the command line.
  -RN2, --read-names-2  As --read-names-1, except the files will show the
                        correspondence between read names and which unique
                        sequence they correspond to. This option cannot be
                        used with either of the --merging-threshold or
                        --excision-coords options, because they change the
                        correspondence initially established between unique
                        sequences and reads.
  --exact-window-start  Experimental; for bioinformatic investigation only,
                        not regular phyloscanner usage. Normally phyloscanner
                        retrieves all reads that fully overlap a given window,
                        i.e. starting at or anywhere before the window start,
                        and ending at or anywhere after the window end. With
                        this option, the reads that are retrieved are those
                        that start at exactly the start of the window, and end
                        anywhere. Window end coordinates are ignored. If
                        combined with --exact-window-end, for a read to be
                        kept it must start at exactly the window start AND end
                        at exactly the window end. If --merge-paired-reads is
                        also used, this explanation applies to inserts (read
                        pairs) instead of individual reads.
  --exact-window-end    With this option, the reads that are retrieved are
                        those that end at exactly the end of the window, and
                        start anywhere. Read the --exact-window-start help.
  -CE, --recover-clipped-ends
                        The default behaviour of phyloscanner is to keep only
                        reads that fully span the window in question. A read
                        which is long enough in principle to reach the edge of
                        the window but is not mapped at its end, i.e. the end
                        is clipped, will therefore not be included. With this
                        option, clipped ends are recovered by considering any
                        bases at the ends of the read that are unmapped to be
                        mapped instead to 1 more than the base to their left
                        (at the right end) or 1 less than the base to their
                        right (at the left end), iterating out from the
                        centre. e.g. a 9bp read mapped to positions
                        None,None,10,11,13,14,None,None,None (i.e. clipped on
                        the left by 2bp, and on the right by 3bp, with a 1bp
                        deletion in the middle), is taken to be mapped instead
                        to positions 8,9,10,11,13,14,15,16,17. In this
                        example, if the window left edge is 8 or 9 and the
                        right edge is 15, 16 or 17, the read with its clipped
                        ends recovered spans the window but the read without
                        clipped ends does not. WARNING: mapping software clips
                        the ends of reads for a reason, namely that that
                        stretch of sequence does not look anything like the
                        reference at that point. The clipped sequence could be
                        just junk, or genuine sample from a distant part of
                        the genome (i.e. the read is chimeric); in this case
                        the clipped sequence should be discarded. As such,
                        this option should not be used as part of normal
                        phyloscanner usage. Its intended usage is specifically
                        the following: you have identified a window in a bam
                        file in which reads are clipped, but you believe the
                        reads to be correct, i.e. the clipping is an artefact
                        of the mapper being unable to find the correct local
                        alignment. You should combine this option with --no-
                        trees because the inclusion of clipped sequence, which
                        by definition is very different, increases the chance
                        of misalignment. You should inspect the aligned reads
                        manually before doing anything else (and hopefully get
                        some insight into how the reference in this window
                        should be changed in order to have subsequent
                        remapping get the local alignment right, in particular
                        by contrasting the reference with the consensus of the
                        aligned reads).

Options to only partially run phyloscanner, stopping early or skipping steps:
  -AO, --align-refs-only
                        Align the mapping references used to create the bam
                        files, plus any extra reference sequences specified
                        with -A, then quit without doing anything else. The
                        point is to allow inspection of this alignment, whose
                        coordinates are used to interpret window coordinates.
  -CO, --flag-contaminants-only
                        For each window, just flag contaminant reads then move
                        on (without aligning reads or making a tree). Only
                        makes sense with the -C flag.
  -RNO, --read-names-only
                        To be combined with --read-names-1 or --read-names-2:
                        quit after writing the read names to a file (which
                        means the reads are not aligned).
  -T, --no-trees        Process and align the reads from each window, then
                        quit without making trees.
  -D, --dont-check-duplicates
                        Don't compare reads between samples to find duplicates
                        - a possible indication of contamination. (By default
                        this check is done.)
```
