\documentclass{article}
\usepackage{graphicx}
\usepackage[a4paper, total={16cm, 24cm}]{geometry}
\usepackage{xspace}
\usepackage[hidelinks]{hyperref}
\usepackage{color}
\newcommand{\p}{\texttt{phyloscanner}\xspace}
\newcommand{\pmt}{\texttt{phyloscanner\_make\_trees.py}\xspace}
\newcommand{\pat}{\texttt{phyloscanner\_analyse\_trees.R}\xspace}
\newcommand{\pcl}{\texttt{phyloscanner\_clean\_alignment.R}\xspace}
\let\c\texttt
\newcommand{\www}{\color{blue} \underline}
\newcommand{\R}{\c{RAxML}\xspace}
\newcommand{\iq}{\c{IQtree}\xspace}

\title{\p\\available from \href{https://github.com/BDI-pathogens/phyloscanner}{\www{github.com/BDI-pathogens/phyloscanner}}}
\date{This manual last updated \today}
\author{Coding lead by Chris Wymant and Matthew Hall,\\with contributions from Oliver Ratmann and Christophe Fraser}

\begin{document}
\maketitle


\begin{figure}[!h]
\centering
\includegraphics[width=0.95\textwidth]{PhyloscannerDiagram_big4.pdf}
\end{figure}

\vspace*{5mm}

\p analyses pathogen genetic diversity and relationships between and within hosts at once, in windows along the genome using mapped {\it reads} (fragments of DNA produced by next-generation sequencing), or using phylogenies with multiple sequences per host previously generated by the user via any method.
An example showing its action on simulated data (which is included with the code) can be seen at the GitHub repository linked to above.

\vspace*{5mm}

If you have a query about \p, ask it publicly at the \href{https://groups.google.com/forum/\#!forum/phyloscanner-users}{\www{google group}}.  
If you have a problem getting the code to run or think you've found a bug, please create a \href{https://github.com/BDI-pathogens/phyloscanner/issues}{\www{\it New issue}} on the GitHub repository.
The usual bug-reporting etiquette applies.
For non-trivial bugs, to allow us to reproduce the problem it's helpful to have input that causes the problem.
This needn't be all your data.
For example if one read in a bam file is causing a problem, you could make a new bam file containing only that read, allowing us to see the problem very quickly.
Isolate the problem as much as you can first -- help us to help you!
If you get an error when you run \p as part of a larger piece of code you've written yourself (e.g. a job submission script to a cluster), please don't make us debug that -- the error might be in your code rather than ours.
Isolate the problem to the relevant \p command.

\vspace*{5mm}

If you use \p, please cite it~\cite{Wymant157768} and its dependencies: \c{SAMtools}~\cite{Li08062009}, \c{MAFFT}~\cite{Katoh15072002}, \R~\cite{doi:10.1093/bioinformatics/btu033} and \c{ggtree}~\cite{MEE3:MEE312628}.
Details of these are found in the file \c{CitationDetails.bib} in \p's \c{InfoAndInputs} subdirectory.

\vspace*{5mm}

Throughout the manual we'll use {\it phylogeny} and {\it tree} interchangeably.
Forgive us.

\newpage
\tableofcontents
\newpage

\section*{A tool in two parts}
Conceptually, \p can be neatly split into two steps.
Firstly, inferring phylogenies that contain both within- and between-host diversity, in windows along the genome, using mapped reads as input.
Secondly, analysing those phylogenies or other similar ones (i.e. containing within- and between-host diversity, but created however the user likes) in order to infer interesting things like transmission and multiply infected individuals, and give summaries of diversity.

This conceptual split is reflected in the code: \pmt runs the first step and \pat runs the second step.
Why didn't we merge the two steps into one?
\begin{itemize}
\item For some uses, step one may not be needed.
For example if you don't have mapped reads, but have sequence data in another format.
We presented an example of this in the \p paper: a dataset consisting of multiple whole genomes of {\it Streptococcus pneumoniae} for each person carrying the bacterium, obtained by separately sequencing multiple colonies.
\item For some uses, step two may not be needed.
Having extracted the within- and between-host diversity information in each window (first as a multiple sequence alignment and then as a phylogeny), you might want to do your own totally different analysis of it.
%Step one consists of extracting the right reads from all samples for each window, processing them, aligning them, then running \R on the aligned reads.
\item Each step is controlled by parameters and options to tailor its usage to different kinds of data, and we don't (yet) have a way to automatically optimise these for each data set, so you will need to have a little play to explore what's appropriate in your case.
If you are running both steps, it's best to think first about getting step one right first, and then getting step two right.
Step one is primarily about bioinformatics, step two about phylogenetic analysis.
\end{itemize}

Note that to run either \pmt or \pat, like running any executable file, you must either specify the full path to the executable (e.g.\\\c{/path/to/my/phyloscanner/code/}\pmt) or add the directory that contains the executable to the \c{\$PATH} environment variable in your terminal (google this if needed).
In the manual we'll refer to scripts named like \c{tools/SomeCode}; such scripts live in the \c{tools} subdirectory inside the main \p code directory.


Fun fact: step one was written by Chris, step two mostly by Matthew.
We sat next to each other while coding, and we hope that's reflected in how easy it is to make the two steps work together.
If it isn't, let us know.
%While the two steps can be run separately, we had their combined use  


\include{PhyloscannerManual_Part1}
\include{PhyloscannerManual_Part2}

\bibliographystyle{ieeetr}
\bibliography{CitationDetails}

\end{document}
