\documentclass{article}
\usepackage{graphicx}
\usepackage[a4paper, total={16cm, 24cm}]{geometry}
\usepackage{xspace}
\usepackage{hyperref}
\newcommand{\p}{\texttt{phyloscanner}\xspace}
\let\c\texttt

\title{\p\\available from \href{https://github.com/BDI-pathogens/phyloscanner}{github.com/BDI-pathogens/phyloscanner}}
\date{This manual last updated \today}
\author{Coding lead by Chris Wymant and Matthew Hall,\\with contributions from Oliver Ratmann and Christophe Fraser}

\begin{document}
\maketitle


\begin{figure}[!h]
\centering
\includegraphics[width=0.95\textwidth]{PhyloscannerDiagram_big4.pdf}
\end{figure}

\vspace*{10mm}

\p analyses pathogen genetic diversity and relationships between and within hosts at once, in windows along the genome using mapped {\it reads} (fragments of DNA produced by next-generation sequencing), or using phylogenies with multiple sequences per host previously generated by the user via any method.

\vspace*{10mm}

If you have a problem getting the code to run or think you've found a bug, please create a \href{https://github.com/BDI-pathogens/phyloscanner/issues}{{\it New issue}} on the GitHub repository.
The usual bug-reporting etiquette applies.
(1) For non-trivial bugs, to allow us to reproduce the problem it's helpful to have input files that cause the problem.
This needn't be all your data.
For example if one read in a bam file is causing a problem, you could make a new bam file containing only that read, allowing us to see the problem very quickly.
Isolate the problem as much as you can first -- help us to help you!
(2) If you get an error when you run \p as part of a larger piece of code you've written yourself (e.g. a job submission script to a cluster), please don't make us debug that -- the error might be in your code rather than ours.
Isolate the problem to the relevant \p command.  


\include{PhyloscannerManual_Part1}
\include{PhyloscannerManual_Part2}


\end{document}
